util/Pos/Util/Verification.hs

```diff
@@ -0,0 +1,96 @@
+{- |
+
+Machinery to support pure datatype verification.
+
+1. Datatypes can have and not have internal pure consistency. If
+datatype doesn't have any implicit rules, it's called safe. 'Integer'
+is safe, 'Either Text String' is safe too. Otherwise, it's called
+unsafe. 'Coin' is unsafe, because it has implicit upper bound on the
+value, which is lower than 'Word64' bound. Newtype over string that is
+supposed to contain only strings of length less than 30 symbols is
+unsafe.
+
+2. Unsafe types' constructors should have prefix \"Unsafe\" (for
+example, 'UnsafeProxySecretKey'). They should also have description of
+why are they unsafe and provide some easy-to-use builders (@mkX@) if
+needed.
+
+3. If datatype @S@ is a part of @T@ (e.g. record field), then @T@
+is not called unsafe.
+
+4. @T@ is called verifiable if it is unsafe or any of its parts are
+verifiable.
+
+5. Every verifiable type must have 'PVerifiable' instance. See
+description of this class.
+
+-}
```

Comment:
I saw that somewhere you wrote why this module is useful. I think you should also write it here, because at first glance it looks like unnecessary overcomplication.

Reply:
Ok, will add.